File: C++/minimum-index-sum-of-two-lists.cpp
// Time:  O(m + n), m is the size of list1, n is the size of list2
// Space: O(m * l), l is the average string length

class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        unordered_map<string, int> lookup;
        for (int i = 0; i < list1.size(); ++i) {
            lookup[list1[i]] = i;
        }
        vector<string> result;
        int min_sum = numeric_limits<int>::max();
        for (int j = 0; j < list2.size() && j <= min_sum; ++j) {
            if (lookup.count(list2[j])) {
                auto sum = j + lookup[list2[j]];
                if (sum < min_sum) {
                    result.clear();
                    result.emplace_back(list2[j]);
                    min_sum = sum;
                } else if (sum == min_sum)
                    result.emplace_back(list2[j]);
            }
        }
        return result;
    }
};

File: Python/minimum-index-sum-of-two-lists.py
# Time:  O((m + n) * l), m is the size of list1, n is the size of list2
# Space: O(m * l), l is the average length of string

class Solution(object):
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        lookup = {}
        for i, s in enumerate(list1):
            lookup[s] = i

        result = []
        min_sum = float("inf")
        for j, s in enumerate(list2):
            if j > min_sum:
                break
            if s in lookup:
                if j + lookup[s] < min_sum:
                    result = [s]
                    min_sum = j + lookup[s]
                elif j + lookup[s] == min_sum:
                    result.append(s)
        return result

